File: back/src/main/java/com/back/domain/concert/entity/Concert.java
package com.back.domain.concert.entity;

import com.back.global.jpa.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Concert extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long concertId;

    @Column(nullable = false)
    private String concertName;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private LocalDateTime startDate;

    @Column(nullable = false)
    private LocalDateTime endDate;

    private String urlPoster;

    private Concert(String concertName, String description, LocalDateTime startDate, LocalDateTime endDate, String urlPoster) {
        this.concertName = concertName;
        this.description = description;
        this.startDate = startDate;
        this.endDate = endDate;
        this.urlPoster = urlPoster;
    }

    public static Concert create(String concertName, String description, LocalDateTime startDate, LocalDateTime endDate, String urlPoster) {
        return new Concert(concertName, description, startDate, endDate, urlPoster);
    }
}

File: back/src/main/kotlin/com/back/domain/concert/entity/Concert.kt
package com.back.domain.concert.entity

import com.back.global.jpa.entity.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import java.time.LocalDateTime

@Entity
class Concert(
    concertName: String,
    description: String? = null,
    startDate: LocalDateTime,
    endDate: LocalDateTime,
    urlPoster: String? = null
) : BaseEntity() {

    @Column(nullable = false)
    var concertName: String = concertName
        protected set

    @Column(columnDefinition = "TEXT")
    var description: String? = description
        protected set

    @Column(nullable = false)
    var startDate: LocalDateTime = startDate
        protected set

    @Column(nullable = false)
    var endDate: LocalDateTime = endDate
        protected set

    var urlPoster: String? = urlPoster
        protected set

    // AI가 여러 관람후기를 종합해 생성한 요약. REVIEW가 3건 이상 모이기 전까지는 null이다.
    @Column(length = 1000)
    var reviewSummary: String? = null
        protected set

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val concertId: Long? = null

    // 저장 후에는 항상 not-null인 PK. 조회된(영속화된) Concert에서만 사용해야 한다.
    val concertIdOrThrow: Long
        get() = checkNotNull(concertId) { "concertId must not be null after persist" }

    fun isBookable(): Boolean = endDate.isAfter(LocalDateTime.now())

    fun applyReviewSummary(summary: String) {
        this.reviewSummary = summary
    }

    companion object {
        fun create(
            concertName: String,
            description: String?,
            startDate: LocalDateTime,
            endDate: LocalDateTime,
            urlPoster: String?
        ): Concert = Concert(
            concertName = concertName,
            description = description,
            startDate = startDate,
            endDate = endDate,
            urlPoster = urlPoster
        )
    }
}
